polkadot/consensus/src/service.rs

```diff
@@ -0,0 +1,232 @@
+
+	fn start_send(&mut self, message: bft::Communication) -> ::futures::StartSend<bft::Communication, E> {
+		let network_message = match message {
+			bft::generic::Communication::Consensus(c) => net::BftMessage::Consensus(match c {
+				bft::generic::LocalizedMessage::Propose(proposal) => net::SignedConsensusMessage::Propose(net::SignedConsensusProposal {
+					round_number: proposal.round_number as u32,
+					proposal: proposal.proposal,
+					digest: proposal.digest,
+					sender: proposal.sender,
+					digest_signature: proposal.digest_signature.signature,
+					full_signature: proposal.full_signature.signature,
+				}),
+				bft::generic::LocalizedMessage::Vote(vote) => net::SignedConsensusMessage::Vote(net::SignedConsensusVote {
+					sender: vote.sender,
+					signature: vote.signature.signature,
+					vote: match vote.vote {
+						bft::generic::Vote::Prepare(r, h) => net::ConsensusVote::Prepare(r as u32, h),
+						bft::generic::Vote::Commit(r, h) => net::ConsensusVote::Commit(r as u32, h),
+						bft::generic::Vote::AdvanceRound(r) => net::ConsensusVote::AdvanceRound(r as u32),
+					}
+				}),
+			}),
+			bft::generic::Communication::Auxiliary(justification) => net::BftMessage::Auxiliary(justification.uncheck().into()),
+		};
+		self.network.send_bft_message(network_message);
+		Ok(::futures::AsyncSink::Ready)
+	}
+
+	fn poll_complete(&mut self) -> ::futures::Poll<(), E> {
+		Ok(Async::Ready(()))
+	}
+}
+
+/// Consensus service. Starts working when created.
+pub struct Service {
+	thread: Option<thread::JoinHandle<()>>,
+}
+
+struct Network(Arc<net::ConsensusService>);
+
+impl Service {
+	/// Create and start a new instance.
+	pub fn new<C>(client: Arc<C>, network: Arc<net::ConsensusService>, transaction_pool: Arc<Mutex<TransactionPool>>, best_header: &Header) -> Service
+		where C: BlockchainEvents + bft::BlockImport + bft::Authorities + PolkadotApi + Send + Sync + 'static
+	{
+		let best_header = best_header.clone();
+		let thread = thread::spawn(move || {
+			let mut core = reactor::Core::new().expect("tokio::Core could not be created");
+			let key = Arc::new(Keyring::One.into());
+			let factory = ProposerFactory {
+				client: client.clone(),
+				transaction_pool: transaction_pool.clone(),
+				network: Network(network.clone()),
+			};
+			let bft_service = BftService::new(client.clone(), key, factory);
+			// Kickstart BFT agreement on start.
+			if let Err(e) = Self::run_bft(&bft_service, network.clone(), &*client, core.handle(), &best_header) {
+				debug!("Error starting initial BFT agreement: {:?}", e);
+			}
+			loop {
+				let handle = core.handle();
+				let start_bft = client.import_notification_stream().map(|notification| {
+					if let Err(e) = Self::run_bft(&bft_service, network.clone(), &*client, handle.clone(), &notification.header) {
+						debug!("Error starting BFT agreement: {:?}", e);
+					}
+				}).map_err(|e| debug!("BFT agreement error: {:?}", e));
+				if let Err(_e) = core.run(start_bft.into_future()) {
```

Comment:
instead of doing this in a loop, couldn't we do

`let bft_starter = client.import_notification_stream().for_each(|notification| ...)` and `core.run(bft_starter)`?